File: src_fortran/sub/nittoutput.f
c
c
c
c
c
c
      subroutine NITTOUTPUT(damp)
c
c     output the results of iteration NITT:
c
      implicit none
      real damp
      include '../inc/vel_com.inc'
c
      real avdt,avdx,avdy,avdz,aavdt,aavdx,aavdy,aavdz
      real avelo
      integer i,j2,j1,j,k,k2,j11,j22,kj,jjj,ifl,ksta1,k1,kk1
      integer ksta2
      real cc(ist)
      character*1 reflch
c
c     output hypocenters of this iteration:
c
      avdt=0.0
      avdx=0.0
      avdy=0.0
      avdz=0.0
      aavdt=0.0
      aavdx=0.0
      aavdy=0.0
      aavdz=0.0
      write(16,*)
      write(16,'(4h  eq, 7x, 2hot, 5x, 1hx, 6x, 1hy, 6x, 1hz, 6x,3hrms,
     &           4x,5havres,''   dot     dx     dy     dz'' )')
c
      do 29 i=1,legs
c
c     print constrain-info if necessary:
c
      if(isingle.ne.0)then
         if(isconstrain(1).eq.1)then
            write(16,*)' *** nitt<2 --> ',
     &                 'depth-adjustment := 0.0 for event ',i
         endif
         if(isconstrain(2).eq.1)then
            write(16,*)' *** igap>250 --> ',
     &                 'depth-adjustment := 0.0 for event ',i
         endif
         if(isconstrain(3).eq.1)then
            write(16,*)' *** depth-adjustment constrained for event ',i
         endif
      endif
      if(iconstrain(i).eq.1)then
         write(16,'('' ***** depth constrained for event '',i5)') i
      endif
c
c   output new hypocenter-results got in this iteration:
c
      j2=4*i
      j1=j2-3
      if(i.gt.neqs)then  ! for shots
         j1=3*neqs+i
         j2=j1
      endif
      if(icoordsystem.eq.2)then
         write(16,37) i,e(1,i),-e(2,i),(e(j,i),j=3,4),rms(i),avres(i),
     &             b(j1), -b(j1+1), +b(j1+2), b(j1+3)      ! dot  dx  dy  dz
      else
         write(16,37) i,(e(j,i),j=1,4),rms(i),avres(i),
     &                (b(j),j=j1,j2)              ! dot  dx  dy  dz
      endif
 37   format (1x,i4,3x,6f7.2,2x,4f7.3)
      avdt=avdt+b(j1)
      avdx=avdx+b(j1+1)
      avdy=avdy+b(j1+2)
      avdz=avdz+b(j1+3)
      aavdt=aavdt+ABS(b(j1))
      aavdx=aavdx+ABS(b(j1+1))
      aavdy=aavdy+ABS(b(j1+2))
      aavdz=aavdz+ABS(b(j1+3))
c
  29  continue  ! loop over all events (printout)
c
      if(icoordsystem.eq.2) avdx=-avdx
      avdt=avdt/float(legs)
      avdx=avdx/float(legs)
      avdy=avdy/float(legs)
      avdz=avdz/float(legs)
      aavdt=aavdt/float(legs)
      aavdx=aavdx/float(legs)
      aavdy=aavdy/float(legs)
      aavdz=aavdz/float(legs)
      write(16,*)
      write(16,'('' A V E R A G E   of ADJUSTMENTS :'',19x,4f7.3)')
     &           avdt,avdx,avdy,avdz
      write(16,'('' A V E R A G E   of ABSOLUTE ADJUSTMENTS :'',10x,
     &           4f7.3)') aavdt,aavdx,aavdy,aavdz
      write(16,*)
c
c     all new hypocenters printed out now.
c
      if(damp.ne.1.0)then
         write(16,124) damp
124      format(/,' Step length damping of ',f7.5,' was applied.',/)
      else
         write(16,'(/,''NO step length damping applied'',/)')
      endif
c
c     print velocity adjustments and velocity-model here:
c
      if(scale(6).eq.0.0) goto 44  ! no velocity-adjustments...
      j1=4*neqs+nshot+1
      j2=j1+nltot-1
      k=0
      write(16,*)
      write(16,40)
 40   format(' Velocity adjustments:')
      write(16,41)
 41   format(5x,'vp    dvp      hp   reflector')
c
c     do it for all velocity-models
c
      do i=1,nmod
         write(16,71) i
71       format(1x,'Velocity model',i4)
         k2=laysum(i)
         j11=j1+k2-1
         j22=j11+nplay(i)-1
         kj=1
c
c        do it for all layers in this model ( = i )
c
         do jjj=j11,j22
            if(kj.eq.ireflector)reflch=reflchar
            if(kj.ne.ireflector)reflch=' '
            write(16,63) vp(i,kj),b(jjj),hp(i,kj),reflch
 63         format(1x,2f7.3,2x,f7.3,5x,a1)
            kj=kj+1
         enddo
c
c        calculate and print average velocities of the model i :
c
         ifl=1
         write(16,*)
         write(16,*)'Calculation of average velocity starts at layer #',
     &              ifl
         avelo=0
         do kj=ifl+1,nplay(i)
            avelo=avelo + ( hp(i,kj)-hp(i,kj-1) ) * vp(i,kj-1)
            write(16,'('' z ='',f6.2,'' ...'',f6.2,''    v = '',f4.2,
     &                ''    AVEL = '',f4.2,'' downto z = '',f6.2)')
     &           hp(i,kj-1),hp(i,kj),vp(i,kj-1),
     &           avelo/(hp(i,kj)-hp(i,ifl)),hp(i,kj)
         enddo
         write(16,*)
c
         ifl=2
         write(16,*)
         write(16,*)'Calculation of average velocity starts at layer #',
     &              ifl
         avelo=0
         do kj=ifl+1,nplay(i)
            avelo=avelo + ( hp(i,kj)-hp(i,kj-1) ) * vp(i,kj-1)
            write(16,'('' z ='',f6.2,'' ...'',f6.2,''    v = '',f4.2,
     &                ''    AVEL = '',f4.2,'' downto z = '',f6.2)')
     &           hp(i,kj-1),hp(i,kj),vp(i,kj-1),
     &           avelo/(hp(i,kj)-hp(i,ifl)),hp(i,kj)
         enddo
         write(16,*)
         write(16,*)
      enddo  ! loop over all models
c
c     each velocity-model and its changes printed out now.
c
c     print p- & s- station-corrections here:
c
 44   if(scale(5).eq.0.0) goto 72 ! no station-correction adjustments...
      write(16,*)
      write(16,'('' Adjusted station corrections:'')')
      write(16,48)
 48   format(2x,' stn  ptcor  dpcor  ')
      k1=4*neqs+nshot+nltot+1
      ksta1=ksta
      if(nsp.eq.2) ksta1=ksta/2
      k2=k1+ksta1-1
c
c     print p-correction for all stations:
c
      do j=1,nsta
         cc(j)=0.
         if(map1(j).eq.0) goto 49
         if(map1(j).gt.ksta1) goto 49
         kk1=k1-1+map1(j)
         cc(j)=b(kk1)             !  STATION-CORRECTION-
cccc         ptcor(j)=ptcor(j)+cc(j)  !  ADJUSTMENT !!! (p-correction)
 49      continue
      enddo
      write(16,50) (stn(j),ptcor(j),cc(j),j=1,nsta)
 50   format(4(2x,a4,2f7.3))
      if(nsp.ne.2) goto 72
      write(16,'('' Adjusted station corrections:'')')
      write(16,73)
73    format(2x,' stn  stcor  dscor  ')
      k1=4*neqs+nshot+nltot+1+ksta1
      ksta2=ksta-ksta1
      k2=k1+ksta2-1
c
c     print s-correction for all stations:
c
      do j=1,nsta
         cc(j)=0.
         if(map1(j).eq.0) goto 74
         if(map1(j).gt.ksta2) goto 74
         kk1=k1-1+map1(j)
         cc(j)=b(kk1)              ! STATION-CORRECTION-
ccc         stcor(j)=stcor(j)+cc(j)   ! ADJUSTMENT !!! (s-correction)
74       continue
      enddo
      write(16,50) (stn(j),stcor(j),cc(j),j=1,nsta)
c
72    continue
      write(16,*)
c
      RETURN
c
      end ! of subr. nittoutput

File: src/sub/nittoutput.c
#include <math.h>
#include <stdio.h>
#include <stdbool.h>
#include "../include/globals.h"

extern int legs;
extern int neqs;
extern int nshot;
extern int nltot;
extern int nmod;
extern int nsp;
extern int nsta;
extern int ksta;
extern int icoordsystem;
extern bool single_turbo;
extern int isingle;
extern int ireflector;
extern char reflchar;
extern FILE *fm_ptr;

extern int isconstrain[];
extern int iconstrain[];
extern int map1[];
extern double e[5][IEQ];
extern double b[];
extern double rms[IEQ];
extern double avres[IEQ];
extern double scale[];
extern double vp[ITOTMODELS][INLTOT];
extern double hp[ITOTMODELS][INLTOT];
extern int nplay[INLTOT];
extern int laysum[INLTOT];
extern double ptcor[IST];
extern double stcor[IST];
extern char stn[IST][5];

void nittoutput(double damp) {
    double avdt = 0.0, avdx = 0.0, avdy = 0.0, avdz = 0.0;
    double aavdt = 0.0, aavdx = 0.0, aavdy = 0.0, aavdz = 0.0;
    double cc[IST] = {0.0};

    FILE *out = fm_ptr ? fm_ptr : stdout;

    fprintf(out, "\n");
    fprintf(out, "  eq       ot     x      y      z      rms   avres   dot     dx     dy     dz\n");

    for (int i = 0; i < legs; ++i) {
        if (isingle != 0) {
            if (isconstrain[0] == 1) fprintf(out, " *** nitt<2 --> depth-adjustment := 0.0 for event %d\n", i + 1);
            if (isconstrain[1] == 1) fprintf(out, " *** igap>250 --> depth-adjustment := 0.0 for event %d\n", i + 1);
            if (isconstrain[2] == 1) fprintf(out, " *** depth-adjustment constrained for event %d\n", i + 1);
        }
        if (iconstrain[i] == 1) {
            fprintf(out, " ***** depth constrained for event %5d\n", i + 1);
        }

        int j1 = 4 * i;
        if (i >= neqs) {
            j1 = 3 * neqs + i;
        }

        if (icoordsystem == 2) {
                fprintf(out, " %4d %7.2f%7.2f%7.2f%7.2f%7.2f%7.2f %7.3f%7.3f%7.3f%7.3f\n",
                    i + 1, e[0][i], -e[1][i], e[2][i], e[3][i], rms[i], avres[i],
                    b[j1], -b[j1 + 1], b[j1 + 2], b[j1 + 3]);
        } else {
                fprintf(out, " %4d %7.2f%7.2f%7.2f%7.2f%7.2f%7.2f %7.3f%7.3f%7.3f%7.3f\n",
                    i + 1, e[0][i], e[1][i], e[2][i], e[3][i], rms[i], avres[i],
                    b[j1], b[j1 + 1], b[j1 + 2], b[j1 + 3]);
        }

        avdt += b[j1];
        avdx += b[j1 + 1];
        avdy += b[j1 + 2];
        avdz += b[j1 + 3];
        aavdt += fabs(b[j1]);
        aavdx += fabs(b[j1 + 1]);
        aavdy += fabs(b[j1 + 2]);
        aavdz += fabs(b[j1 + 3]);
    }

    if (icoordsystem == 2) avdx = -avdx;

    avdt /= (double)legs;
    avdx /= (double)legs;
    avdy /= (double)legs;
    avdz /= (double)legs;
    aavdt /= (double)legs;
    aavdx /= (double)legs;
    aavdy /= (double)legs;
    aavdz /= (double)legs;

        fprintf(out, "\n A V E R A G E   of ADJUSTMENTS :                    %7.3f%7.3f%7.3f%7.3f\n",
            avdt, avdx, avdy, avdz);
        fprintf(out, " A V E R A G E   of ABSOLUTE ADJUSTMENTS :           %7.3f%7.3f%7.3f%7.3f\n\n",
            aavdt, aavdx, aavdy, aavdz);

    if (damp != 1.0f) {
        fprintf(out, "\n Step length damping of %7.5f was applied.\n\n", damp);
    } else {
        fprintf(out, "\nNO step length damping applied\n\n");
    }

    if (scale[4] != 0.0) {
        int j1 = 4 * neqs + nshot + nltot;
        int ksta1 = (nsp == 2) ? (ksta / 2) : ksta;

        fprintf(out, "\n Adjusted station corrections:\n");
        fprintf(out, "  stn  ptcor  dpcor\n");

        for (int j = 0; j < nsta; ++j) {
            cc[j] = 0.0;
            if (map1[j] > 0 && map1[j] <= ksta1) {
                int kk1 = j1 + map1[j] - 1;
                cc[j] = b[kk1];
            }
            fprintf(out, "  %4s %7.3f %7.3f\n", stn[j], ptcor[j], cc[j]);
        }

        if (nsp == 2) {
            int ksta2 = ksta - ksta1;
            j1 = 4 * neqs + nshot + nltot + ksta1;
            fprintf(out, " Adjusted station corrections:\n");
            fprintf(out, "  stn  stcor  dscor\n");
            for (int j = 0; j < nsta; ++j) {
                cc[j] = 0.0;
                if (map1[j] > 0 && map1[j] <= ksta2) {
                    int kk1 = j1 + map1[j] - 1;
                    cc[j] = b[kk1];
                }
                fprintf(out, "  %4s %7.3f %7.3f\n", stn[j], stcor[j], cc[j]);
            }
        }
    }

    fprintf(out, "\n");
}
